core/src/main/java/org/apache/iceberg/avro/TypeToSchema.java

```diff
@@ -238,8 +218,68 @@ public Schema primitive(Type.PrimitiveType primitive) {
         throw new UnsupportedOperationException("Unsupported type ID: " + primitive.typeId());
     }
 
-    results.put(primitive, primitiveSchema);
-
     return primitiveSchema;
   }
+
+  static class WithTypeToName extends TypeToSchema {
+
+    private final Map<Type, Schema> results = Maps.newHashMap();
+
+    WithTypeToName(Map<Types.StructType, String> names) {
+      super((id, struct) -> names.get(struct));
+    }
+
+    Map<Type, Schema> getConversionMap() {
+      return results;
+    }
+
+    @Override
+    public Schema struct(Types.StructType struct, List<Schema> fieldSchemas) {
+      Schema recordSchema = results.get(struct);
+      if (recordSchema != null) {
+        return recordSchema;
+      }
+
+      recordSchema = super.struct(struct, fieldSchemas);
+      results.put(struct, recordSchema);
+      return recordSchema;
+    }
+
+    @Override
+    public Schema list(Types.ListType list, Schema elementSchema) {
+      Schema listSchema = results.get(list);
+      if (listSchema != null) {
+        return listSchema;
+      }
+
+      listSchema = super.list(list, elementSchema);
+      results.put(list, listSchema);
+      return listSchema;
+    }
+
+    @Override
+    public Schema map(Types.MapType map, Schema keySchema, Schema valueSchema) {
+      Schema mapSchema = results.get(map);
+      if (mapSchema != null) {
+        return mapSchema;
+      }
+
+      mapSchema = super.map(map, keySchema, valueSchema);
+      results.put(map, mapSchema);
+      return mapSchema;
+    }
+
+    @Override
+    public Schema primitive(Type.PrimitiveType primitive) {
+      Schema primitiveSchema = super.primitive(primitive);
+      results.put(primitive, primitiveSchema);
+      return primitiveSchema;
+    }
+  }
+
+  static class WithNamesFunction extends TypeToSchema {
+    WithNamesFunction(BiFunction<Integer, Types.StructType, String> namesFunction) {
```

Comment:
Should this also cache to dedup, but by type name rather than by structure? It may not be urgent, but I'm a little concerned about failing Avro conversion because we try to create two schemas with the same name. I think that's what we were trying to avoid with the cache in the first place.

It isn't really a risk for existing code because this class is only used by a new method, though.

Reply:
I went ahead and added the caching by record name also.